synapse/storage/engines/_base.py

```diff
@@ -133,8 +133,9 @@ def executescript(cursor: CursorType, script: str) -> None:
 
         This is not provided by DBAPI2, and so needs engine-specific support.
 
-        Some database engines may automatically COMMIT the ongoing transaction both
-        before and after executing the script.
+        Some database engines may automatically COMMIT the ongoing transaction before
+        executing the script in its own transaction. The script transaction is left
+        open and it is the responsibility of the caller to commit it.
```

Comment:
> Some database engines may automatically COMMIT the ongoing transaction

I wonder if we should make this a hard guarantee of this function, so that we have the same semantics across postgres and sqlite.

For example, suppose we need to apply two migrations A and B to upgrade the database schema. IIUC, under postgres, synapse will run

```sql
BEGIN TRANSACTION;
...
<A>;
<B>;
COMMIT;
```

whereas under sqlite we will now run
```sql
BEGIN TRANSACTION;
...
COMMIT; BEGIN TRANSACTION;
<A>;
COMMIT; BEGIN TRANSACTION;
<B>;
COMMIT;
```

I'm not sure if that's a difference we should worry about. WDYT?

Reply:
I'm in favour of the idea. It's much easier to reason about things when both database engines behave the same way.

I gave that a go in bb340f7f0a0db4d21b49e98ed4847075a461e59e. Let's hope it passes CI.